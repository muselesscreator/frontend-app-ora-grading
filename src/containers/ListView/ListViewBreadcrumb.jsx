import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { ArrowBack, Launch } from '@edx/paragon/icons';
import { Hyperlink, Icon } from '@edx/paragon';

import selectors from 'data/selectors';
import { locationId } from 'data/constants/app';
import urls from 'data/services/lms/urls';

/**
 * <ListViewBreadcrumb />
 */
export const ListViewBreadcrumb = ({ courseId, oraName }) => (
  <>
    <Hyperlink className="py-4" destination={urls.openResponse(courseId)}>
      <ArrowBack className="mr-3" />
      Back to all open responses
    </Hyperlink>
    <p className="h3 py-4">
      {oraName}
      <Hyperlink destination={urls.ora(courseId, locationId)} target="_blank">
        <Icon icon={Launch} />
      </Hyperlink>
    </p>
  </>
);
ListViewBreadcrumb.defaultProps = {
  courseId: '',
  oraName: '',
};
ListViewBreadcrumb.propTypes = {
  courseId: PropTypes.string,
  oraName: PropTypes.string,
};

export const mapStateToProps = (state) => ({
  courseId: selectors.app.courseId(state),
  oraName: selectors.app.ora.name(state),
});

export const mapDispatchToProps = {};

export default connect(mapStateToProps, mapDispatchToProps)(ListViewBreadcrumb);
