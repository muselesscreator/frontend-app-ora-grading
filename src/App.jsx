import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

import Footer from '@edx/frontend-component-footer';

import { selectors } from 'data/redux';

import ListView from 'containers/ListView';
import './App.scss';

import Header from 'containers/CourseHeader';
import DemoBanner from './components/DemoBanner';

export const App = ({ courseMetadata }) => (
  <Router>
    <div>
      <Header
        courseTitle={courseMetadata.title}
        courseNumber={courseMetadata.number}
        courseOrg={courseMetadata.org}
      />
      <DemoBanner />
      <main>
        <ListView />
      </main>
      <Footer logo={process.env.LOGO_POWERED_BY_OPEN_EDX_URL_SVG} />
    </div>
  </Router>
);
App.defaultProps = {
  courseMetadata: {
    title: '',
    number: null,
    org: '',
  },
};
App.propTypes = {
  courseMetadata: PropTypes.shape({
    title: PropTypes.string,
    number: PropTypes.string,
    org: PropTypes.string,
  }),
};

export const mapStateToProps = (state) => ({
  courseMetadata: selectors.app.courseMetadata(state),
});

export default connect(mapStateToProps)(App);
