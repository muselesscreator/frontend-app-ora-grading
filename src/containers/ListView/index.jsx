import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
  DataTable,
  TextFilter,
  MultiSelectDropdownFilter,
  Container,
} from '@edx/paragon';

import {
  gradingStatusDisplay,
} from 'data/services/lms/constants';

import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';

import StatusBadge from 'components/StatusBadge';
import ReviewModal from 'containers/ReviewModal';
import ListViewBreadcrumb from './ListViewBreadcrumb';
import TableControls from './TableControls';
import './ListView.scss';

export const gradeStatusOptions = Object.keys(gradingStatusDisplay).map(key => ({
  name: gradingStatusDisplay[key],
  value: key,
}));

/**
 * <ListView />
 */
export class ListView extends React.Component {
  constructor(props) {
    super(props);
    this.props.initializeApp();
    this.handleViewAllResponsesClick = this.handleViewAllResponsesClick.bind(this);
    this.selectedBulkAction = this.selectedBulkAction.bind(this);
  }

  formatDate = ({ value }) => {
    const date = new Date(value);
    return date.toLocaleString();
  }

  formatGrade = ({ value: grade }) => (
    grade === null ? '-' : `${grade.pointsEarned}/${grade.pointsPossible}`
  );

  formatStatus = ({ value }) => (<StatusBadge status={value} />);

  handleViewAllResponsesClick(data) {
    const getSubmissionId = (row) => row.original.submissionId;
    const rows = data.selectedRows.length ? data.selectedRows : data.tableInstance.rows;
    this.props.loadSelectionForReview(rows.map(getSubmissionId));
  }

  selectedBulkAction(selectedFlatRows) {
    return {
      buttonText: `View selected responses (${selectedFlatRows.length})`,
      className: 'view-selected-responses-btn',
      handleClick: this.handleViewAllResponsesClick,
      variant: 'primary',
    };
  }

  render() {
    // hide if submissions are not loaded.
    if (this.props.listData.length === 0) {
      return null;
    }

    return (
      <Container className="py-4">
        <ListViewBreadcrumb />
        <DataTable
          isFilterable
          numBreakoutFilters={2}
          defaultColumnValues={{ Filter: TextFilter }}
          isSelectable
          isSortable
          isPaginated
          itemCount={this.props.listData.length}
          initialState={{ pageSize: 10, pageIndex: 0 }}
          data={this.props.listData}
          tableActions={[
            {
              buttonText: 'View all responses',
              handleClick: this.handleViewAllResponsesClick,
              className: 'view-all-responses-btn',
              variant: 'primary',
            },
          ]}
          bulkActions={[
            this.selectedBulkAction,
          ]}
          columns={[
            {
              Header: 'Username',
              accessor: 'username',
            },
            {
              Header: 'Learner submission date',
              accessor: 'dateSubmitted',
              Cell: this.formatDate,
              disableFilters: true,
            },
            {
              Header: 'Grade',
              accessor: 'score',
              Cell: this.formatGrade,
              disableFilters: true,
            },
            {
              Header: 'Grading Status',
              accessor: 'gradingStatus',
              Cell: this.formatStatus,
              Filter: MultiSelectDropdownFilter,
              filter: 'includesValue',
              filterChoices: gradeStatusOptions,
            },
          ]}
        >
          <TableControls />
        </DataTable>
        <ReviewModal />
      </Container>
    );
  }
}
ListView.defaultProps = {
  listData: [],
};
ListView.propTypes = {
  initializeApp: PropTypes.func.isRequired,
  listData: PropTypes.arrayOf(PropTypes.shape({
    username: PropTypes.string,
    dateSubmitted: PropTypes.number,
    gradingStatus: PropTypes.string,
    grade: PropTypes.shape({
      pointsEarned: PropTypes.number,
      pointsPossible: PropTypes.number,
    }),
  })),
  loadSelectionForReview: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  listData: selectors.submissions.listData(state),
});

export const mapDispatchToProps = {
  initializeApp: thunkActions.app.initialize,
  loadSelectionForReview: thunkActions.grading.loadSelectionForReview,
};

export default connect(mapStateToProps, mapDispatchToProps)(ListView);
