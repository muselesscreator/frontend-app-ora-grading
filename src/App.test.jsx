import React from 'react';
import { shallow } from 'enzyme';

import Footer from '@edx/frontend-component-footer';

import ListView from 'containers/ListView';

import { App } from './App';

jest.mock('@edx/frontend-component-footer', () => 'Footer');
jest.mock('components/DemoBanner', () => 'DemoBanner');
jest.mock('containers/ListView', () => 'ListView');
jest.mock('containers/CourseHeader', () => 'CourseHeader');

const logo = 'fakeLogo.png';
let el;
let router;

describe('App router component', () => {
  const props = {
    courseMetadata: {
      org: 'course-org',
      number: 'course-number',
      title: 'course-title',
    },
  };
  test('snapshot', () => {
    expect(shallow(<App {...props} />)).toMatchSnapshot();
  });
  describe('component', () => {
    beforeEach(() => {
      process.env.LOGO_POWERED_BY_OPEN_EDX_URL_SVG = logo;
      el = shallow(<App {...props} />);
      router = el.childAt(0);
    });
    describe('Router', () => {
      test('Routing - ListView is only route', () => {
        expect(router.find('main')).toEqual(shallow(
          <main><ListView /></main>,
        ));
      });
    });
    test('Footer logo drawn from env variable', () => {
      expect(router.find(Footer).props().logo).toEqual(logo);
    });
  });
});
