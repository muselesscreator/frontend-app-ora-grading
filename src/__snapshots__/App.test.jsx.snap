// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App router component snapshot 1`] = `
<BrowserRouter>
  <div>
    <CourseHeader
      courseNumber="course-number"
      courseOrg="course-org"
      courseTitle="course-title"
    />
    <DemoBanner />
    <main>
      <ListView />
    </main>
    <Footer
      logo="https://edx-cdn.org/v3/stage/open-edx-tag.svg"
    />
  </div>
</BrowserRouter>
`;
